### ISSUE TITLE
quimb.tensor.tensor_1d - implement OTOC

### ISSUE THREAD

**Issue by TanTsiChen** [jcmgray/quimb#7]
It's known from AdS/CFT that the out-of-time-ordered correlators (OTOC) is considered as a measure of
quantum chaos. In SYK model the OTOC gives the Lyapunov boundary (boundary of quantum Lyapunov exponent), recently it's pointed out that the second Rényi entropy is related to OTOC (see ref. [1](https://linkinghub.elsevier.com/retrieve/pii/S2095927317301925)). It is interesting that if an OTOC generating by the order parameters of the model, then this OTOC can represent the dynamical phase transition in the model (a non-equilibrium process), and even give the quantum critical points (see ref. [2](https://doi.org/10.1103/PhysRevLett.121.016801)). So it's interesting to implement OTOC in quimb.
Consider we have a spin chain, one can see that the OTOC generating by a pauli operator or spin operator is an important case, in this issues i will push my option to calculate the OTOC which is generated by a local pauli 'z' notation.

**Comment by TanTsiChen** [comment]
In fact, the OTOC on Transverse-field Ising chain is a very good example, and the results can be compared with the results in ref. [2](https://doi.org/10.1103/PhysRevLett.121.016801). Let's begin with the repeating the results in this paper.
First, we need to crate a up-state (with PBC) like in this paper, for this purpose, we need to do a little modification (add the PBC case) to the following two functions: `MPS_product_state(arrays, **mps_opts)` and `MPS_computational_state(binary, dtype=float, **mps_opts)`.

`def MPS_product_state(arrays, cyclic=Flase, **mps_opts):
    """Generate a product state in MatrixProductState form with PBC or OBC,
    i,e, with bond dimension 1, from single site vectors described by ``arrays``.
    """
    cyc_dim = (1,) if cyclic else ()

    def gen_array_shapes():
        yield (*cyc_dim, 1, -1)
        for _ in range(len(arrays) - 2):
            yield (1, 1, -1)
        yield (*cyc_dim, 1, -1)

    mps_arrays = (_asarray(array).reshape(*shape)
                  for array, shape in zip(arrays, gen_array_shapes()))

    return MatrixProductState(mps_arrays, shape='lrp', **mps_opts)`

`def MPS_computational_state(binary, dtype=float, cyclic=False, **mps_opts):
    """A computational basis state in Matrix Product State form.

    Parameters
    ----------
    binary : str or sequence of int
        String specifying the state, e.g. ``'00101010111'`` or ``[0, 0, 1]``.
    cyclic : bool, optional
        Generate a MPS with periodic boundary conditions or not, default open boundary conditions.
    mps_opts
        Supplied to MatrixProductState constructor.
    """
    cyclic = cyclic
    array_map = {
        '0': np.array([1., 0.], dtype=dtype),
        '1': np.array([0., 1.], dtype=dtype),
    }

    def gen_arrays():
        for s in binary:
            yield array_map[str(s)]

    return MPS_product_state(tuple(gen_arrays()), cyclic, **mps_opts)`

Now, the up-state with PBC can be given as following
`up_chain = '0' * L
psi0 = MPS_computational_state(up_chain, cyclic=True)`

then we generate a MPO of local pauli 'z' notation as following
`def MPO_local_pauli_z(L, i, phys_dim=2, cyclic=False, phase=False, **mpo_opts):
    """generate a MPO of the local pauli 'z' notation acting on site 'i'
    
    Parameters
    ----------
    
    L : int
        The length of chain.
    i : int
        The site where the local pauli 'z' notation acting on.
    phys_dim : int, optional
        The physical (site) dimensions, defaults to 2.
    cyclic : bool, optional
        Generate a MPO with periodic boundary conditions or not, default is open
        boundary conditions.
    phase : bool, optional
        Add a phase shift e^{i\pi/2} or not, default is no phase shift.
    mpo_opts
        Supplied to :class: '~quimb.tensor.tensor_1d.MatrixProductOperator'.
    """
    
    
    II = np.identity(phys_dim, dtype=complex)
       
    if phase is False:
        z = np.array([[1., 0.], [0., -1.]], dtype=complex)
    else:
        z = np.array([[0.+1.j, 0.], [0., 0.-1.j]], dtype=complex)
    
    cyc_dim = (1,) if cyclic else ()
    
    
    def gen_array():
        if i == 1:
            yield z.reshape(*cyc_dim, 1, phys_dim, phys_dim)
        else:
            yield II.reshape(*cyc_dim, 1, phys_dim, phys_dim)        
        for n in range(L-2):
            if i == n+2:
                yield z.reshape(1, 1, phys_dim, phys_dim)
            else:
                yield II.reshape(1, 1, phys_dim, phys_dim)
        if i == L:
            yield z.reshape(1, *cyc_dim, phys_dim, phys_dim)
        else:
            yield II.reshape(1, *cyc_dim, phys_dim, phys_dim)
        
    return qtn.MatrixProductOperator(gen_array(), **mpo_opts)
`
Here the reason why we add a 'phase' option to 'MPO_local_pauli_z' is that in OTOC calculation we need to apply the i \sigma^z_n operation (which with a phase shift i \pi/2) to the MPS.

**Comment by TanTsiChen** [comment]
Then we can calculate the OTOC

###### def pauli_z_OTOC(L, i, t, psi0, H, H_back, dt=None, tol=None, order=4, split_opts=None):
    """ The out-of-time-ordered correlation (OTOC) function generating by a
    local pauli 'z' notation acting on site 'i' at time 't', it's a function
    of time.
    
    Parameters
    ----------
    L : int
        The length of chain.
    i : int
        The site where the local pauli 'z' notation acting on.
    t : float
        The time to evolve to.
    psi0 : MatrixProductState
        The initial state in MPS form.
    H : NNI
        The Hamiltonian for forward time-evolution.
    H_back : NNI
        The Hamiltonian for backward time-evolution, should have only sign difference
        with 'H'.
    dt : float, optional
        Time step to use, cannot be set as well as tol.
    tol : float, optional
        Tolerance for whole evolution, cannot be set as well as dt.
    order : int, optional
        Trotter order to use, defaults to 4.
    split_opts : dict, optional
        Compression options applied for splitting after gate application, see
        tensor_split().
    """
    
    # forward time evolution
    tebd1 = qtn.TEBD(psi0, H, split_opts=split_opts)
    tebd1.update_to(t, dt=dt, tol=tol, order=order)
    
    # local pauli 'z' notation with a phase shift applys to tebd1.pt
    i_sigma_z = MPO_local_pauli_z(L, i, phase=True)
    psit_sigma = i_sigma_z.apply(tebd1.pt)

    # backward time evolution
    tebd2 = qtn.TEBD(psit_sigma, H_back, split_opts=split_opts)
    tebd2.update_to(t, dt=dt, tol=tol, order=order)

    # return the expective value of local pauli 'z'
    return qtn.expec_TN_1D(tebd2.pt.H, MPO_local_pauli_z(L, i), tebd2.pt)

**Comment by TanTsiChen** [comment]
Then we just need to calculate OTOC for up transverse-field Ising chain as following
`H = qtn.NNI_ham_ising(L, j=4, bx=1, cyclic=True)`
`H_back = qtn.NNI_ham_ising(L, j=-4, bx=-1, cyclic=True)`
`F = pauli_z_OTOC(L, i, t, psi0, H, H_back, dt=1e-2)`
&
`H = qtn.NNI_ham_ising(L, j=4, bx=3, cyclic=True)`
`H_back = qtn.NNI_ham_ising(L, j=-4, bx=-3, cyclic=True)`
`F = pauli_z_OTOC(L, i, t, psi0, H, H_back, dt=1e-2)`
and check the result by comparing with the result in [paper](https://doi.org/10.1103/PhysRevLett.121.016801)

**Comment by jcmgray** [comment]
@TanTsiChen thanks for the contribution. In order for me to review this and (possibly merge it) you will have to create a pull request - then I can checkout and test the code properly.

If you haven't done this before, that roughly involves:
1. Forking ``quimb`` and making sure its up-to-date.
2. Making your additions and committing then pushing them to *your* fork on github.
3. Opening a [PR](https://github.com/jcmgray/quimb/pulls) against my repo.

There is a [very good guide here](http://xarray.pydata.org/en/latest/contributing.html). It's for the [``xarray``](https://github.com/pydata/xarray/) project but all the information translates pretty well (to almost all open-source projects including this one).

**Comment by jcmgray** [comment]
@TanTsiChen it look's like you were on the right lines. I haven't studied OTOCs but from inspecting your code does this function work for you?:
```python
import quimb as qu
import quimb.tensor as qtn


def OTOC(psi0, H, H_back, t, i, phased=False):
    
    # obviously play with these settings
    tebd_opts = {
        'tol': 1e-5,
        'split_opts': {
            'cutoff': 1e-5,
            'cutoff_mode': 'rel',
        },
    }
    
    Z = qu.pauli('Z')
    if phased:
        Z = 1j * Z
    
    # evolve forward
    tebd1 = qtn.TEBD(psi0, H, **tebd_opts)
    tebd1.update_to(t)
    
    # apply first Z-gate - note you don't need an MPO
    psi_t_Z = tebd1.pt.gate(Z, i, contract=True)
    
    # evolve backwards
    tebd2 = qtn.TEBD(psi_t_Z, H_back, **tebd_opts)
    tebd2.update_to(t)
    
    # compute expectation with second Z-gate
    psi_f = tebd2.pt
    return psi_f.H.expec(psi_f.gate(Z, i, contract=True))
```

**Comment by TanTsiChen** [comment]
It's a little different from your code, phased is not a option. The calculation procedure is as the following:
1. evolve forward
2. apply a phased Z-gate
3. evolve backwards
4. compute expectation with non-phased Z-gate

**Comment by jcmgray** [comment]
Ah OK I have updated the function.

**Comment by TanTsiChen** [comment]
> Ah OK I have updated the function.

so, you decide to put this function into which module?

**Comment by jcmgray** [comment]
Closed by #10
